README
NAME
    Module::Version - Get module versions

VERSION
    Version 0.11

SYNOPSIS
    This module fetches the version of any other module.

    It comes with a CLI program "mversion" which does the same.

        use Module::Version 'get_version';

        print get_version('Search::GIN'), "\n";

    Or using "mversion":

        $ mversion Search::GIN
        0.04

        $ mversion Doesnt::Exist
        Warning: module 'Doesnt::Exist' does not seem to be installed.

        $ mversion --quiet Doesnt::Exist
        (no output)

        $ mversion --full Search::GIN Moose
        Search::GIN 0.04
        Moose 1.01

        $ mversion --input modules.txt
        Search::GIN 0.04
        Data::Collector 0.03
        Moose 1.01

EXPORT
  get_version
    "get_version" will be exported if explicitly specified.

        use Module::Version 'get_version';

    Nothing is exported by default.

SUBROUTINES/METHODS
  get_version
    Accepts a module name and fetches the version of the module.

    If the module doesn't exist, returns undef.

AUTHOR
    Sawyer X, "<xsawyerx at cpan.org>"

BUGS
    Please report bugs and other issues on the bugtracker:

    <http://github.com/xsawyerx/module-version/issues>

SUPPORT
    This module sports 100% test coverage, but in case you have more issues,
    please see *BUGS* above.

LICENSE AND COPYRIGHT
    Copyright 2010 Sawyer X.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
{{$NEXT}}

0.201     2019-02-24 22:58:48+02:00 Asia/Jerusalem

        * Fix tests.
        * Generate minimum Perl version. (Mohammad S Anwar)

0.200     2018-12-20 00:06:37+02:00 Asia/Jerusalem

        * Add support for local::lib.
          (Yanick Champoux)
        * Require Test::Output for the testing.
          (Paul Cochrane)
        * Add initial Travis CI support.
          (Paul Cochrane)
        * Fix test failure relating to incorrect hard-coded version.
          (Yanick Champoux)

0.13      2018-08-26 23:55:36+03:00 Asia/Jerusalem

        * Fix a test failure when Test::More's VERSION ends with unnecessary
          zeros. Thanks, Andreas J. König.
        * Cleanups.
        * Dzilify.

0.12    29.06.10
        * Missed another one :(

0.11    29.06.10
        * Fixing some more documentation I missed.

0.10    28.06.10
        * Adding -I option, GH #1.
        * Using cleanup regex for dev versions instead of eval.
          (otherwise, versions like 0.10 don't work)
        * Fixing some documentations to refer to new CLI app name.

0.09    20.06.10
        * CLI app now called "mversion" instead of "module-version".
        * Using autodie, it's fun!
        * 100% testing coverage! (I was bored...)

0.08    25.05.10
        * Improving documentation of module-version program.

0.07    25.05.10
        * Documenting module-version program.

0.06    24.05.10
        * Adding --dev (RT #57795):
          (using --dev you get 0.01_01 on dev instead of 0.0101)

0.05    23.04.10
        * Using high minimum version of ExtUtils::MakeMaker to avoid test
          errors.

0.04    21.04.10
        * Switched to use ExtUtils::MakeMaker instead of evaling.
        (it's now _way_ faster than before)

0.03    10.04.10
        * Docs fix.
        * Better test coverage.

0.02    10.04.10
        * Bumping version for test script.
        * Adding Module::Version::App.
        * Adding module-version script.

0.01    10.04.10
        First version, released on an unsuspecting world.



lib/Module/Version.pm
package Module::Version;

use strict;
use warnings;

use base 'Exporter';
use Carp;
use ExtUtils::MakeMaker;

our $VERSION   = '0.11';
our @EXPORT_OK = 'get_version';

sub get_version {
    my $module = shift or croak 'Must get a module name';
    my $file   = MM->_installed_file_for_module($module);

    $file || return;

    return MM->parse_version($file)
}

1;

__END__

=head1 NAME

Module::Version - Get module versions

=head1 VERSION

Version 0.11

=head1 SYNOPSIS

This module fetches the version of any other module.

It comes with a CLI program C<mversion> which does the same.

    use Module::Version 'get_version';

    print get_version('Search::GIN'), "\n";

Or using C<mversion>:

    $ mversion Search::GIN
    0.04

    $ mversion Doesnt::Exist
    Warning: module 'Doesnt::Exist' does not seem to be installed.

    $ mversion --quiet Doesnt::Exist
    (no output)

    $ mversion --full Search::GIN Moose
    Search::GIN 0.04
    Moose 1.01

    $ mversion --input modules.txt
    Search::GIN 0.04
    Data::Collector 0.03
    Moose 1.01

=head1 EXPORT

=head2 get_version

C<get_version> will be exported if explicitly specified.

    use Module::Version 'get_version';

B<Nothing> is exported by default.

=head1 SUBROUTINES/METHODS

=head2 get_version

Accepts a module name and fetches the version of the module.

If the module doesn't exist, returns undef.

=head1 AUTHOR

Sawyer X, C<< <xsawyerx at cpan.org> >>

=head1 BUGS

Please report bugs and other issues on the bugtracker:

L<http://github.com/xsawyerx/module-version/issues>

=head1 SUPPORT

This module sports 100% test coverage, but in case you have more issues, please
see I<BUGS> above.

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Sawyer X.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



lib/Module/Version/App.pm
package Module::Version::App;
# ABSTRACT: Application implementation for Module::Version

use strict;
use warnings;
use Carp qw< croak >;

use Getopt::Long    qw( :config no_ignore_case );
use Module::Version 'get_version';

our $VERSION = '0.13';

sub new { return bless {}, $_[0] }

sub run {
    my $self    = shift;
    my @modules;

    $self->parse_opts;

    if( $self->{local_lib} and -d  $self->{local_lib} ) {
        require local::lib;
        local::lib->import( $self->{local_lib} );
    }

    $self->{'modules'}
        and push @modules, @{ $self->{'modules'} };

    if ( my $file = $self->{'input'} ) {
        open my $fh, '<', $file
            or croak("Cannot open '$file': $!");

        chomp( my @extra_modules = <$fh> );
        push @modules, @extra_modules;

        close $fh
            or croak("Cannot close '$file': $!");
    }

    if ( $self->{'include'} ) {
        my $include = $self->{'include'};

        ref $include eq 'ARRAY'
            or die "Error: include must be an ARRAY ref\n";

        unshift @INC, @{$include};
    }

    @modules
        or die "Error: no modules to check\n";

    foreach my $module (@modules) {
        my $version = get_version($module);
        if ( !$version ) {
            $self->{'quiet'}
                or warn "Warning: module '$module' does not seem to be installed.\n";

            next;
        }

        $self->{'dev'} or $version =~ s/_(.+)$/$1/xms;

        my $output = $self->{'full'} ? "$module $version\n" : "$version\n";
        print $output;
    }
}

sub parse_opts {
    my $self = shift;

    GetOptions(
        'h|help'       => sub { $self->help },
        'f|full!'      => \$self->{'full'},
        'i|input=s'    => \$self->{'input'},
        'l|local-lib=s'=> \$self->{'local_lib'},
        'I|include=s@' => \$self->{'include'},
        'd|dev!'       => \$self->{'dev'},
        'q|quiet!'     => \$self->{'quiet'},
        '<>'           => sub { $self->process(@_) },
    ) or $self->error('could not parse options');
}

sub process {
    my ( $self, @args ) = @_;

    # force stringify Getopt::Long input
    push @{ $self->{'modules'} }, "$_" for @args;
}

sub help {
    my $self = shift;

    print << "_END_HEREDOC";
$0 [ OPTIONS ] Module Module Module...

Provide a module's version, comfortably.

OPTIONS
    -f | --full      Output name and version (a la Module::Version 0.05)
    -I | --include   Include any number of directories to include as well
    -i | --input     Input file to read module names from 
    -l | --local-lib Additional local::lib dir to search
    -d | --dev       Show developer versions as 0.01_01 instead of 0.0101
    -q | --quiet     Do not error out if module doesn't exist

_END_HEREDOC
}

1;

__END__

=head1 SYNOPSIS

This is the CLI program's implementation as a module.

    use Module::Version::App;

    my $app = Module::Version::App->new;

    $app->run;

=head1 SUBROUTINES/METHODS

=head2 new

Create a new object.

=head2 run

Do all the grunt work.

=head2 parse_opts

Parsing the command line arguments using L<Getopt::Long>.

=head2 process

Parses extra arguments from L<Getopt::Long>.

=head2 help

Print a help menu for the application itself.

=head2 error($error)

Calls C<die> with a message.

=head2 warn($warning)

Calls C<warn> with a message.

=head1 EXPORT

Object Oriented, nothing is exported.

=head1 BUGS

Please report any bugs or feature requests to
C<bug-module-version at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Module-Version>.  I will be
notified, and then you'll automatically be notified of progress on your bug as I
make changes.

=head1 SUPPORT

This module sports 100% test coverage, but in case you have more issues...

You can find documentation for this module with the perldoc command.

    perldoc Module::Version

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Version>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Module-Version>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Module-Version>

=item * Search CPAN

L<http://search.cpan.org/dist/Module-Version/>

=back
